src/main/java/com/google/devtools/build/lib/bazel/bzlmod/Registry.java

```diff
@@ -16,28 +16,50 @@
 package com.google.devtools.build.lib.bazel.bzlmod;
 
 import com.google.common.collect.ImmutableMap;
+import com.google.devtools.build.lib.bazel.repository.downloader.Checksum;
 import com.google.devtools.build.lib.events.ExtendedEventHandler;
 import java.io.IOException;
+import java.net.URI;
 import java.util.Optional;
 
 /** A database where module metadata is stored. */
 public interface Registry {
 
+  @FunctionalInterface
```

Comment:
Overall I think this `Downloader`/`AccessRecorder` mechanism makes the flow rather obscure.

Before, it was pretty simple: `registry.getModuleFile()` gets called, does whatever it needs to retrieve the module file (whether it be HTTP, fake registry stuff, or anything), and returns it.

Now, it needs to be careful to call `recorder.downloadAndRecord()` with the right parameters, _and_ provide a proper `Downloader`. And the _caller_ needs to make sure to provide an `AccessRecorder` that correctly calls the `Downloader` it's given. Control is handed between multiple parties multiple times.

This is reminiscent of the "internal vs external iteration" and the "callback hell vs promises" dichotomy. I would always lean towards the latter choice since it affords composition (irrelevant in our case) and presents a clearer control flow (_relevant_ here).

Question would be how we actually do it. There are two facets to tackle, AIUI:
1) To provide checksums to the registry, I was thinking about some sort of "priming" mechanism for `RegistryFactory` -- as in, a `RegistryFactory` can be "primed" by calling `registryFactory.prime(knownFilesAndChecksums)` (or even just constructed with it). This would require some more refactoring since `RegistryFactory` right now lives outside Skyframe (sort of), but with it we'd always need the lockfile to be an input. I haven't yet figured out what's the best way to structure the code here.

2) To let registry access report checksums, we could use the existing event handlers and have them post events on the event bus. I think that should be enough to let someone somewhere subscribe and produce a good lockfile.

WDYT?

Reply:
Regarding 2., I'm a bit worried about the number of Skyframe events becoming an issue at some point. As far as I understand `storeForReplay`, these events are carried forward to every dependent node. With download events, we could easily get ~1000s of events.

@justinhorvitz Would you expect Skyframe to handle large numbers of replayed events attached to "low-level" nodes in the graph well?

If this is okay, then I agree that rewriting this to go through events is better, which then also unlocks cleaner approaches to do 1.

Reply:
What I specifically meant was [`ExtendedEventHandler.Postable`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/events/ExtendedEventHandler.java;drc=d8d9078ba5dadc575e12f52424313b3a07bf3eaf;l=26) instead of [`Event`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/events/Event.java) (which has the `storeForReplay` stuff), although I'm not too sure about the performance characteristics either way. (Confusingly, `Postable` goes on the `EventBus`, but `Event` doesn't...)

To sidestep this, we could also just let the registry itself remember accesses (essentially have our own `EventBus`, in a way).

Reply:
I extracted an attempt at 1.: https://github.com/bazelbuild/bazel/pull/22040

Instead of bringing `RegistryFactory` into Skyframe, I'm doing it for `Registry`, which seemed simpler. Then `RegistryFunction` can depend on lockfile information when it creates `IndexRegistry`.

Reply:
Firstly, are the posts/events you are considering posted to a [`SkyFunction.Environment#getListener`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/skyframe/SkyFunction.java;l=343;drc=1ca79ac4e996a50f8e33e085112abb7d1f1ec982)? If not, then there's no memory concern and what I'm about to say is [not relevant](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/events/Reportable.java;l=48-49;drc=6f48f1c2b3bb73768b9ff15ce6698e21eddc503a).

`Postable` has `storeForReplay` too, but it [defaults to false](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/events/Reportable.java;l=55;drc=435d1c2707949997380ca1cc13868f14049e4247). If you override that to true, you get the potential memory blowup that you were talking about. If you keep that false, then if the skyframe node is cached on an incremental build, the postable will not be re-posted.

Reply:
@justinhorvitz Yes, we would be both posting them there and returning `true` from `storeForReplay`. 

The events themselves are small and it's not a problem if they are retained. They also use reference equality. As long as the overall memory and CPU cost for the entire build is linear in the number of events regardless of their location in the graph, I guess that we should be fine - but is that the case?

Reply:
> As long as the overall memory and CPU cost for the entire build is linear in the number of events regardless of their location in the graph, I guess that we should be fine - but is that the case?

No - every skyframe node in the reverse transitive closure will use a [`ValueWithMetadata`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/skyframe/ValueWithMetadata.java) wrapper. This is so that regardless of where we prune on an incremental build, the post will be replayed.

Excessive memory use for this is why we did https://github.com/bazelbuild/bazel/commit/08a8f6acd6bddcad8ad167c5e5831b2e5765b83a.

Reply:
Thanks for the clarification, that actually means that we already have a problem: I just checked that with `--enable_bzlmod` and `--lockfile_mode=update`, virtually any Skyframe node in Bazel uses a `ValueWithMetadata` due to a transitive dependency on the nodes emitting `BazelModuleResolutionEvent` and `ModuleExtensionResolutionEvent`. Removing the calls to `ExtendedEventHandler#post` reduces memory usage on Bazel itself from 84 MB to 72 MB (although of course the events themselves also take up memory).

I see some `shouldPropagate` logic that is used in other cases to limit the propagation of events. What properties of a "cut set" would we need to guarantee for limited propagation to be safe incrementally?

Reply:
An alternative: The events are used to update the Bzlmod lockfile in an `afterCommand` function. We started setting `storeForReplay` to `true` to prevent the following situation:
1. Lockfile doesn't exist at the start of the build and is thus tracked as non-existent by Skyframe.
2. The build emits events that provide the new lockfile content, which is written to disk in `afterCommand`. 
3. The user deletes the lockfile and runs another incremental build.
4. From the perspective of Skyframe, nothing has changed as the lockfile doesn't exist now, just as in 1. No events are emitted and the lockfile remains non-existent/incomplete.

I haven't been able to think this through fully, but if we had a way to mark the lockfile as "dirty" in 2, we may not even have to store the events in the first place. @justinhorvitz Do you see a way to achieve this?

Reply:
Tried this, but it doesn't work: Even if the lockfile is marked as dirty, change pruning makes it so that none of the functions firing events are reevaluated.

Reply:
I have a plan:
1. Replace `ModuleExtensionResolutionEvent` with Skyframe walking.
2. Get rid of the `BazelModuleResolutionEvent` as it is not needed with the new lockfile format.
3. Use retained `Postable`s for download events, but ensure that they aren't propagated past the "compute main repo" phase (the main repo mapping would be a natural cutpoint).

@Wyverald What do you think? 

I submitted https://github.com/bazelbuild/bazel/pull/22058 to realize the first step (with one open question).

Reply:
1 and 2 sound great to me. 3 seems a bit too tricky for my taste. I wonder if we could just use a Registry-specific event handling mechanism and not have it meddle with Skyframe at all? Maybe that's not super feasible (I haven't thought this through), but I feel like the tight coupling with Skyframe graph edges could go out of hand very quickly.

Reply:
> I see some shouldPropagate logic that is used in other cases to limit the propagation of events. What properties of a "cut set" would we need to guarantee for limited propagation to be safe incrementally?

In general: phase boundaries (analysis nodes don't need to propagate events to execution nodes). In specific cases, maybe you could have a stricter propagation policy with knowledge of the skyframe graph and when the events are needed. I'm not familiar with the skyframe node types that are being discussed.

I havent digested this whole issue enough to offer alternative solutions.

Reply:
I am no longer using events.